src/org/mozilla/javascript/NativeJSON.java

```diff
@@ -314,11 +335,15 @@ private static Object str(Object key, Scriptable holder,
             return "null";
         }
 
-        if (value instanceof Scriptable && !(value instanceof Callable)) {
-            if (value instanceof NativeArray) {
-                return ja((NativeArray) value, state);
+        if (value instanceof Scriptable) {
+            if (!(value instanceof Callable)) {
+                if (value instanceof NativeArray) {
+                    return ja((NativeArray) value, state);
+                }
+                return jo((Scriptable) value, state);
             }
-            return jo((Scriptable) value, state);
+        } else if (!Undefined.isUndefined(value)) {
```

Comment:
I'm worried that throwing this exception will break backward compatibility. Previously, someone with a complex object who is trying to stringify it might end up with a weird result, and now they will get a less weird result. But if there's a chance that people who were previously getting weird results will now get an exception, then we could easily break a lot of code. Unless the ECMA spec says that we should throw an exception in this case, then do we really need it?

Reply:
The current code sees it as a NativeJavaObject, which is Scriptable, and so it sends it down the path to try to stringify it as a javascript object, because it is not Callable. This is throwing the StackOverflowError mentioned in the linked issue. I believe the ECMA spec says that all Objects that are not Functions should throw a TypeError if they are cyclic or return a JSON string representing an object.

Since I am unwrapping the NativeJavaObject, it is no longer Scriptable. I think at this point it either has to be Undefined or some other Java Object that can't be serialized, so the options are to either throw a more descriptive exception than it had been throwing, return Undefined, or return "`{}`".

I believe prior to https://github.com/mozilla/rhino/commit/4c76ef2b816350a1bc9a553fa5860669e0abfb04 it never made it this far and threw the other exception described in the linked issue:
```
Java class "[Ljava.lang.reflect.Constructor;" has no public instance field or method named "toJSON"
```
That commit let the object make it through to the replacer, but created a worse exception if a replacer did not exist to do something with the object.

I think my TypeError is in line with the historical behavior before the StackOverflowError, but I am open to suggestions. This also mimics the behavior of BigInt, which throws a TypeError when you try to stringify it without a replacer or adding a custom BigInt.prototype.toJSON method.